op-txpool/conditional_txs.go

```diff
@@ -0,0 +1,184 @@
+	invalidAuthenticationErr = &oprpc.JsonError{Message: "invalid authentication", Code: types.TransactionConditionalRejectedErrCode}
+	entrypointSupportErr     = &oprpc.JsonError{Message: "only 4337 Entrypoint contract support", Code: types.TransactionConditionalRejectedErrCode}
+)
+
+type ConditionalTxService struct {
+	log log.Logger
+	cfg *CLIConfig
+
+	limiter             *rate.Limiter
+	backends            map[string]client.RPC
+	entrypointAddresses map[common.Address]bool
+
+	costSummary prometheus.Summary
+	requests    prometheus.Counter
+	failures    *prometheus.CounterVec
+}
+
+func NewConditionalTxService(ctx context.Context, log log.Logger, m metrics.Factory, cfg *CLIConfig) (*ConditionalTxService, error) {
+	backends := map[string]client.RPC{}
+	for _, addr := range cfg.SendRawTransactionConditionalBackends {
+		rpc, err := client.NewRPC(ctx, log, addr)
+		if err != nil {
+			return nil, fmt.Errorf("failed to dial backend %s: %w", addr, err)
+		}
+
+		rpcMetrics := metrics.MakeRPCClientMetrics(addr, m)
+		backends[addr] = client.NewInstrumentedRPC(rpc, &rpcMetrics)
+	}
+
+	limiter := rate.NewLimiter(types.TransactionConditionalMaxCost, int(cfg.SendRawTransactionConditionalRateLimit))
+	entrypointAddresses := map[common.Address]bool{predeploys.EntryPoint_v060Addr: true, predeploys.EntryPoint_v070Addr: true}
+
+	return &ConditionalTxService{
+		log: log,
+		cfg: cfg,
+
+		limiter:             limiter,
+		backends:            backends,
+		entrypointAddresses: entrypointAddresses,
+
+		costSummary: m.NewSummary(prometheus.SummaryOpts{
+			Namespace: MetricsNameSpace,
+			Name:      "txconditional_cost",
+			Help:      "summary of cost observed by *accepted* conditional txs",
+		}),
+		requests: m.NewCounter(prometheus.CounterOpts{
+			Namespace: MetricsNameSpace,
+			Name:      "txconditional_requests",
+			Help:      "number of conditional transaction requests",
+		}),
+		failures: m.NewCounterVec(prometheus.CounterOpts{
+			Namespace: MetricsNameSpace,
+			Name:      "txconditional_failures",
+			Help:      "number of conditional transaction failures",
+		}, []string{"err"}),
+	}, nil
+}
+
+func (s *ConditionalTxService) SendRawTransactionConditional(ctx context.Context, txBytes hexutil.Bytes, cond *types.TransactionConditional) (common.Hash, error) {
+	s.requests.Inc()
+	if !s.cfg.SendRawTransactionConditionalEnabled {
+		s.failures.WithLabelValues("disabled").Inc()
+		return common.Hash{}, endpointDisabledErr
+	}
+	if cond == nil {
+		s.failures.WithLabelValues("missing conditional").Inc()
+		return common.Hash{}, missingConditionalErr
+	}
+
+	// Authenticate the request
+	peerInfo := rpc.PeerInfoFromContext(ctx)
+	authHeader := peerInfo.HTTP.Header.Get(authHeaderKey)
+	if authHeader == "" {
+		s.failures.WithLabelValues("missing auth").Inc()
+		return common.Hash{}, invalidAuthenticationErr
+	}
+	authElems := strings.Split(authHeader, ":")
```

Comment:
Implementing this under `SendRawTransactionConditional` means it cannot easily be used for any other RPC endpoints. I feel like it should be more generic and not coupled this this RPC endpoint